picoquic/logger.c

```diff
@@ -1983,29 +1982,205 @@ void picoquic_txtlog_message_v(picoquic_quic_t* quic, const picoquic_connection_
+        if (dcid != NULL) {
+            fprintf(F, ", DCID ");
+            picoquic_log_connection_id(F, dcid);
+        }
+        fprintf(F, ", reason: %s\n", trigger);
+    }
+}
+
+void textlog_negotiated_alpn(picoquic_cnx_t* cnx, int is_local,
+    uint8_t const* sni, size_t sni_len, uint8_t const* alpn, size_t alpn_len,
+    const ptls_iovec_t* alpn_list, size_t alpn_count)
+{
+#ifdef _WINDOWS
+    UNREFERENCED_PARAMETER(sni);
+    UNREFERENCED_PARAMETER(sni_len);
+    UNREFERENCED_PARAMETER(alpn);
+    UNREFERENCED_PARAMETER(alpn_len);
+    UNREFERENCED_PARAMETER(alpn_list);
+    UNREFERENCED_PARAMETER(alpn_count);
+#endif
+    if (cnx->quic->F_log != NULL && picoquic_cnx_is_still_logging(cnx)) {
+        /* TODO: alpn */
+        picoquic_textlog_negotiated_alpn(cnx->quic->F_log, cnx, 
+            (is_local) ? 0 : 1, 1, alpn_list, alpn_count);
+    }
+}
+
+
+void textlog_transport_extension(picoquic_cnx_t* cnx, int is_local,
+    size_t param_length, uint8_t* params)
+{
+    if (cnx->quic->F_log != NULL && picoquic_cnx_is_still_logging(cnx)) {
+        /* TODO: alpn */
+        picoquic_textlog_transport_extension(cnx->quic->F_log, cnx, (is_local)?0:1, 1, params, param_length);
     }
+}
 
-    if (cnx->f_binlog != NULL) {
-        va_list args;
-        va_start(args, fmt);
-        picoquic_binlog_message_v(cnx, fmt, args);
-        va_end(args);
+void textlog_tls_ticket(picoquic_cnx_t* cnx, uint8_t* ticket, uint16_t ticket_length)
+{
+    if (cnx->quic->F_log != NULL && picoquic_cnx_is_still_logging(cnx)) {
+        picoquic_log_picotls_ticket(cnx->quic->F_log, picoquic_get_logging_cnxid(cnx),
+            ticket, ticket_length);
     }
 }
+
+void textlog_new_connection(picoquic_cnx_t* cnx)
+{
+#ifdef _WINDOWS
+    UNREFERENCED_PARAMETER(cnx);
+#endif
+}
+
+void textlog_close_connection(picoquic_cnx_t* cnx)
+{
+#ifdef _WINDOWS
+    UNREFERENCED_PARAMETER(cnx);
+#endif
+}
+
+void textlog_cc_dump(picoquic_cnx_t* cnx, uint64_t current_time)
+{
+    if (cnx->quic->F_log != NULL && picoquic_cnx_is_still_logging(cnx)) {
+        picoquic_log_congestion_state(cnx->quic->F_log, cnx, current_time);
+    }
+}
+
+struct st_picoquic_unified_login_t textlog_functions = {
+    /* Per context log function */
+    txtlog_context_free_app_message,
+    textlog_quic_pdu,
+    /* Per connection functions */
+    textlog_app_message,
+    textlog_pdu_ex,
+    textlog_packet,
+    textlog_dropped_packet,
+    textlog_buffered_packet,
+    textlog_outgoing_packet,
+    textlog_packet_lost,
+    textlog_negotiated_alpn,
+    textlog_transport_extension,
+    textlog_tls_ticket,
+    textlog_new_connection,
+    textlog_close_connection,
+    textlog_cc_dump
+};
+
+int picoquic_set_textlog(picoquic_quic_t* quic, char const* textlog_file)
+{
+    int ret = 0;
+    FILE* F_log;
+
+    if (quic->F_log != NULL && quic->should_close_log) {
+        (void)picoquic_file_close(quic->F_log);
```

Comment:
The idea was to use it as `quic->F_log = picoquic_file_close(quic->F_log);`

Reply:
Good catch. Fixing that.